### DISCUSSION TITLE
Why is it necessary / beneficial to dynamically load a remote module?

### DISCUSSION THREAD

**Discussion by msaroy-crey** [module-federation/module-federation-examples#2042] [Q&A] [answered]
Hello,

I'm trying to set up shared redux through the redux injection example, but I can't really get around to understand why the host application needs dynamicFederation function to load the remote modules. I referring to [this bootstrap file](https://github.com/module-federation/module-federation-examples/blob/f196c14fb91476152e0c3029e79a1758c7399ffb/redux-reducer-injection/app1/src/bootstrap.js) in host. 

Wouldn't `app2`'s webpack config `exposes` and `shared` scripts make `app2`s exposed script and shared npm packages available? 
Why do I need to / have to load it this way?

Thanks!!

**Comment by ScriptedAlchemy** [comment, marked as answer]
Shared modules are async be default. So you need a dynamic import before you try and static import a stared module otherwise webpack can't pause the app to negotiate dependencies

**Reply by msaroy-crey** [reply]
Thank you! It clears a lot.

What's still missing is how can I get a reference to the container. `window[scope]` is undefined for me. 
My project does not use monorepo (if that matters). As I searched this repo for help [I found this method ](https://github.com/module-federation/module-federation-examples/blob/master/dynamic-system-host/app1/src/utils/getOrLoadRemote.js), and now I'm able to get something for container by providing remoteEntry.js location. 

It kind of seems like black magic though... 😄

**Reply by ScriptedAlchemy** [reply]
The remote container has to be loaded into the page, if its loaded into the page then window[remoteName] will be an object with get, init. 

Webpack injects it for you if you use the @ syntax, otherwise you need to append it to the head of the document manually, wait for it to load, then you can use get/init manually.

**Comment by kmylo** [comment]
could we do the load of the modules(not the build time,but runtime), fetching from a remote config, out of the box without write some document head script loader tool?

**Reply by ScriptedAlchemy** [reply]
Yeah look at the utils in nextjs-MF. I have an inject script function that uses webpacks own script management system under the hood.

**Reply by ScriptedAlchemy** [reply]
https://github.com/module-federation/nextjs-mf/blob/main/packages/utilities/src/utils/common.ts

**Reply by wibed** [reply]
i am confused.

the `dynamicFederation` mentioned at the beginning seems to be a step back into the beginnings.
from my perspective this has been replaced by the expose, declare module and lazy load routine.

```
Webpack injects it for you if you use the @ syntax, otherwise you need to append it to the head of the document manually, wait for it to load, then you can use get/init manually.
```

can i expose an app including its store and lazy load it and assume the example aforementioned is still archaic.
or is the dynamicFederation step necessary?

tl:dr it feels this dicussion just revamps what module federation is trying to solve, BUT
federating a store might be more inclusive to complexity that I assume it to be.
so here i ask

**Comment by ScriptedAlchemy** [comment]
Adding api to federation runtime to allow registerRemote: https://github.com/module-federation/universe/pull/2171

You should no longer use utilities package as /runtime has all the things you want or need now

https://github.com/module-federation/universe/tree/main/packages/runtime
